File: src/api/v1/endpoints/question_bank.py
"""Question bank endpoints — CRUD + import (URL crawl / document upload)."""

import asyncio
import json
import math
import logging
import os
import tempfile
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from src.core.database import get_db
from src.models.user import User
from src.api.v1.dependencies import get_current_user
from src.schemas.question_bank import (
    QuestionCreate, QuestionUpdate, QuestionResponse,
    QuestionListResponse, QuestionPreview,
    ImportExtractRequest, ImportConfirmRequest, ImportConfirmResponse,
    CATEGORIES, LEVELS,
)
from src.services.data import question_bank_service as svc
from src.services.data.question_importer import extract_from_url, extract_from_file

logger = logging.getLogger(__name__)
router = APIRouter()


def _require_admin(current_user: User = Depends(get_current_user)) -> User:
    if not current_user.is_admin:
        raise HTTPException(status_code=403, detail="Admin access required")
    return current_user


def _to_response(q) -> QuestionResponse:
    return QuestionResponse(
        id=q.id,
        category=q.category,
        subcategory=q.subcategory,
        level=q.level,
        topic=q.topic,
        question=q.question,
        answer=q.answer,
        source=q.source,
        created_at=q.created_at.isoformat(),
        updated_at=q.updated_at.isoformat(),
    )


# ── Meta ──────────────────────────────────────────────────────────────────────

@router.get("/meta")
async def get_meta():
    """Return valid categories and levels for UI dropdowns."""
    return {"categories": CATEGORIES, "levels": LEVELS}


# ── CRUD ──────────────────────────────────────────────────────────────────────

@router.get("", response_model=QuestionListResponse)
async def list_questions(
    category: Optional[str] = Query(None),
    level: Optional[str] = Query(None),
    search: Optional[str] = Query(None),
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(_require_admin),
):
    items, total = await svc.list_questions(db, category, level, search, page, per_page)
    return QuestionListResponse(
        items=[_to_response(q) for q in items],
        total=total,
        page=page,
        per_page=per_page,
        pages=math.ceil(total / per_page) if total else 1,
    )


@router.post("", response_model=QuestionResponse, status_code=201)
async def create_question(
    body: QuestionCreate,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(_require_admin),
):
    q = await svc.create_question(db, body)
    return _to_response(q)


@router.put("/{question_id}", response_model=QuestionResponse)
async def update_question(
    question_id: int,
    body: QuestionUpdate,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(_require_admin),
):
    q = await svc.update_question(db, question_id, body)
    if not q:
        raise HTTPException(status_code=404, detail="Question not found")
    return _to_response(q)


@router.delete("/{question_id}", status_code=204)
async def delete_question(
    question_id: int,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(_require_admin),
):
    deleted = await svc.delete_question(db, question_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Question not found")


# ── Import — shared helpers ────────────────────────────────────────────────────

async def _build_preview(extracted, source: str, db: AsyncSession) -> list[QuestionPreview]:
    previews = []
    for item in extracted:
        status, similar_id, score = await svc.check_duplicate_status(
            db, item.question, item.answer
        )
        previews.append(QuestionPreview(
            category=item.category,
            subcategory=item.subcategory,
            level=item.level,
            topic=item.topic,
            question=item.question,
            answer=item.answer,
            source=source,
            status=status,
            similar_id=similar_id,
            similarity_score=round(score, 4) if score else None,
        ))
    return previews


def _sse(event: dict) -> str:
    return f"data: {json.dumps(event)}\n\n"


# ── Import — extract URL (streaming SSE) ──────────────────────────────────────

@router.post("/import/extract-url/stream")
async def extract_from_url_stream(
    body: ImportExtractRequest,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(_require_admin),
):
    """Stream Playwright crawl + GPT extraction progress via Server-Sent Events."""
    if not body.url:
        raise HTTPException(status_code=400, detail="url is required")

    queue: asyncio.Queue = asyncio.Queue()

    async def log_cb(msg: str) -> None:
        await queue.put({"type": "log", "message": msg})

    _OVERALL_TIMEOUT = 300  # 5 minutes hard ceiling

    async def run() -> None:
        try:
            extracted, source = await asyncio.wait_for(
                extract_from_url(body.url, instructions=body.instructions, log_cb=log_cb),
                timeout=_OVERALL_TIMEOUT,
            )
            if not extracted:
                await queue.put({"type": "error", "message": "No Q&A found on the page"})
                return

            n = len(extracted)
            await log_cb(f"→ Checking {n} question(s) against existing bank (pgvector)…")
            previews = await _build_preview(extracted, source, db)
            new_c = sum(1 for p in previews if p.status == "new")
            sim_c = sum(1 for p in previews if p.status == "similar")
            dup_c = sum(1 for p in previews if p.status == "duplicate")
            await log_cb(f"✓ Done — {new_c} new · {sim_c} similar · {dup_c} duplicate")
            await queue.put({"type": "result", "data": [p.model_dump() for p in previews]})
        except asyncio.TimeoutError:
            await queue.put({"type": "error", "message": f"Extraction timed out after {_OVERALL_TIMEOUT}s. Try a smaller page or fewer menu items."})
        except Exception as e:
            logger.exception("extract_from_url_stream failed")
            await queue.put({"type": "error", "message": str(e)})
        finally:
            await queue.put(None)  # sentinel

    async def generate():
        task = asyncio.create_task(run())
        try:
            while True:
                try:
                    item = await asyncio.wait_for(queue.get(), timeout=_OVERALL_TIMEOUT + 10)
                except asyncio.TimeoutError:
                    yield _sse({"type": "error", "message": "Stream watchdog timeout — server may be overloaded."})
                    break
                if item is None:
                    break
                yield _sse(item)
        finally:
            task.cancel()

    return StreamingResponse(
        generate(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


# ── Import — extract URL (legacy non-streaming) ────────────────────────────────

@router.post("/import/extract-url", response_model=list[QuestionPreview])
async def extract_from_url_endpoint(
    body: ImportExtractRequest,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(_require_admin),
):
    if not body.url:
        raise HTTPException(status_code=400, detail="url is required")
    try:
        extracted, source = await extract_from_url(body.url, instructions=body.instructions)
    except Exception as e:
        raise HTTPException(status_code=422, detail=f"Crawl failed: {e}")
    if not extracted:
        raise HTTPException(status_code=422, detail="No Q&A found on the page")
    return await _build_preview(extracted, source, db)


# ── Import — extract file ──────────────────────────────────────────────────────

@router.post("/import/extract-file", response_model=list[QuestionPreview])
async def extract_from_file_endpoint(
    file: UploadFile = File(...),
    instructions: Optional[str] = Form(None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(_require_admin),
):
    ext = os.path.splitext(file.filename or "")[1].lstrip(".").lower()
    if ext not in ("pdf", "docx", "doc", "txt", "md"):
        raise HTTPException(status_code=400, detail="Supported formats: pdf, docx, txt, md")

    content = await file.read()
    suffix = f".{ext}"
    with tempfile.NamedTemporaryFile(delete=False, suffix=suffix) as tmp:
        tmp.write(content)
        tmp_path = tmp.name

    try:
        extracted, source = await extract_from_file(tmp_path, ext, instructions=instructions)
    except Exception as e:
        raise HTTPException(status_code=422, detail=f"Parsing failed: {e}")
    finally:
        try:
            os.unlink(tmp_path)
        except OSError:
            pass

    if not extracted:
        raise HTTPException(status_code=422, detail="No Q&A found in the document")
    return await _build_preview(extracted, source, db)


# ── Import — confirm ───────────────────────────────────────────────────────────

@router.post("/import/confirm", response_model=ImportConfirmResponse)
async def confirm_import(
    body: ImportConfirmRequest,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(_require_admin),
):
    """Bulk-import the user-selected Q&A items."""
    imported = 0
    skipped = 0
    for item in body.questions:
        if item.status == "duplicate":
            skipped += 1
            continue
        await svc.create_question(db, QuestionCreate(
            category=item.category,
            subcategory=item.subcategory,
            level=item.level,
            topic=item.topic,
            question=item.question,
            answer=item.answer,
            source=item.source,
        ))
        imported += 1
    return ImportConfirmResponse(imported=imported, skipped=skipped)

File: frontend/lib/api/question_bank.ts
import { apiClient } from './client';

export interface Question {
  id: number;
  category: string;
  subcategory: string | null;
  level: string;
  topic: string;
  question: string;
  answer: string;
  source: string | null;
  created_at: string;
  updated_at: string;
}

export interface QuestionListResponse {
  items: Question[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

export interface QuestionPreview {
  category: string;
  subcategory: string | null;
  level: string;
  topic: string;
  question: string;
  answer: string;
  source: string | null;
  status: 'new' | 'similar' | 'duplicate';
  similar_id: number | null;
  similarity_score: number | null;
}

export interface QuestionMeta {
  categories: string[];
  levels: string[];
}

export interface CreateQuestionBody {
  category: string;
  subcategory?: string;
  level: string;
  topic: string;
  question: string;
  answer: string;
  source?: string;
}

export const questionBankApi = {
  getMeta: (): Promise<QuestionMeta> =>
    apiClient.get('/api/v1/admin/questions/meta'),

  list: (params: {
    category?: string;
    level?: string;
    search?: string;
    page?: number;
    per_page?: number;
  }): Promise<QuestionListResponse> => {
    const q = new URLSearchParams();
    if (params.category) q.set('category', params.category);
    if (params.level) q.set('level', params.level);
    if (params.search) q.set('search', params.search);
    if (params.page) q.set('page', String(params.page));
    if (params.per_page) q.set('per_page', String(params.per_page));
    return apiClient.get(`/api/v1/admin/questions?${q.toString()}`);
  },

  create: (body: CreateQuestionBody): Promise<Question> =>
    apiClient.post('/api/v1/admin/questions', body),

  update: (id: number, body: Partial<CreateQuestionBody>): Promise<Question> =>
    apiClient.put(`/api/v1/admin/questions/${id}`, body),

  delete: (id: number): Promise<void> =>
    apiClient.delete(`/api/v1/admin/questions/${id}`),

  extractFromUrl: (url: string, instructions?: string): Promise<QuestionPreview[]> =>
    apiClient.post('/api/v1/admin/questions/import/extract-url', { url, instructions }),

  extractFromUrlStream: (
    url: string,
    instructions?: string,
    onLog?: (msg: string) => void,
  ): Promise<QuestionPreview[]> => {
    const BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8003';
    const token = typeof window !== 'undefined' ? localStorage.getItem('auth_token') : null;
    return new Promise(async (resolve, reject) => {
      let response: Response;
      try {
        response = await fetch(`${BASE_URL}/api/v1/admin/questions/import/extract-url/stream`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            ...(token ? { Authorization: `Bearer ${token}` } : {}),
          },
          body: JSON.stringify({ url, instructions }),
        });
      } catch (e) {
        return reject(new Error('Network error'));
      }
      if (!response.ok) {
        return reject(new Error(`Server error: ${response.status}`));
      }
      const reader = response.body!.getReader();
      const decoder = new TextDecoder();
      let buffer = '';
      try {
        while (true) {
          const { done, value } = await reader.read();
          if (done) break;
          buffer += decoder.decode(value, { stream: true });
          const parts = buffer.split('\n\n');
          buffer = parts.pop() ?? '';
          for (const part of parts) {
            if (!part.startsWith('data: ')) continue;
            let evt: { type: string; message?: string; data?: QuestionPreview[] };
            try { evt = JSON.parse(part.slice(6)); } catch { continue; }
            if (evt.type === 'log') onLog?.(evt.message ?? '');
            else if (evt.type === 'result') { resolve(evt.data ?? []); return; }
            else if (evt.type === 'error') { reject(new Error(evt.message)); return; }
          }
        }
        reject(new Error('Stream ended without result'));
      } catch (e) {
        reject(e instanceof Error ? e : new Error('Stream read error'));
      }
    });
  },

  extractFromFile: (
    file: File,
    instructions?: string,
    onProgress?: (p: number) => void,
  ): Promise<QuestionPreview[]> => {
    const form = new FormData();
    form.append('file', file);
    if (instructions) form.append('instructions', instructions);
    const BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8003';
    const token = typeof window !== 'undefined' ? localStorage.getItem('auth_token') : null;
    return new Promise((resolve, reject) => {
      const xhr = new XMLHttpRequest();
      xhr.open('POST', `${BASE_URL}/api/v1/admin/questions/import/extract-file`);
      if (token) xhr.setRequestHeader('Authorization', `Bearer ${token}`);
      if (onProgress) xhr.upload.onprogress = e => { if (e.lengthComputable) onProgress(Math.round(e.loaded / e.total * 100)); };
      xhr.onload = () => {
        if (xhr.status >= 200 && xhr.status < 300) {
          try { resolve(JSON.parse(xhr.responseText)); } catch { resolve([] as QuestionPreview[]); }
        } else {
          try { reject(new Error(JSON.parse(xhr.responseText).detail || xhr.statusText)); } catch { reject(new Error(xhr.statusText)); }
        }
      };
      xhr.onerror = () => reject(new Error('Network error'));
      xhr.send(form);
    });
  },

  confirmImport: (questions: QuestionPreview[]): Promise<{ imported: number; skipped: number }> =>
    apiClient.post('/api/v1/admin/questions/import/confirm', { questions }),
};
